File: src/features/companies/dto/create-company.dto.ts
import { Transform, Type } from 'class-transformer';
import { IsArray, IsEmail, IsNumber, IsOptional, IsString } from 'class-validator';

export class CreateCompanyDto {
  @IsString()
  @Transform(({ value }) => (typeof value === 'string' ? value.trim() : value))
  name: string;

  @IsOptional()
  @IsString()
  @Transform(({ value }) => (typeof value === 'string' ? value.trim() : value))
  externalId?: string;

  @IsOptional()
  @IsString()
  @Transform(({ value }) => (typeof value === 'string' ? value.trim() : value))
  website?: string;

  @IsOptional()
  @IsEmail()
  @Transform(({ value }) => (typeof value === 'string' ? value.trim().toLowerCase() : value))
  mainEmail?: string;

  @IsOptional()
  @IsString()
  @Transform(({ value }) => (typeof value === 'string' ? value.trim() : value))
  mainPhone?: string;

  @IsOptional()
  @IsArray()
  @IsString({ each: true })
  companyTypeKeys?: string[];

  @IsOptional()
  @IsArray()
  @IsString({ each: true })
  tagKeys?: string[];

  @IsOptional()
  @IsNumber()
  @Type(() => Number)
  rating?: number;

  @IsOptional()
  @IsString()
  @Transform(({ value }) => (typeof value === 'string' ? value.trim() : value))
  notes?: string;
}

File: spring-boot-api/src/main/java/com/mytegroup/api/dto/companies/CreateCompanyDto.java
package com.mytegroup.api.dto.companies;

import jakarta.validation.constraints.Email;
import jakarta.validation.constraints.NotBlank;

import java.util.List;

public record CreateCompanyDto(
    @NotBlank(message = "Name is required")
    String name,
    String externalId,
    String website,
    @Email(message = "Email must be valid")
    String mainEmail,
    String mainPhone,
    List<String> companyTypeKeys,
    List<String> tagKeys,
    Double rating,
    String notes
) {
    public CreateCompanyDto {
        if (name != null) {
            name = name.trim();
        }
        if (externalId != null) {
            externalId = externalId.trim();
        }
        if (website != null) {
            website = website.trim();
        }
        if (mainEmail != null) {
            mainEmail = mainEmail.toLowerCase().trim();
        }
        if (mainPhone != null) {
            mainPhone = mainPhone.trim();
        }
        if (notes != null) {
            notes = notes.trim();
        }
    }
}



